firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/model/ImmutableList.java

```diff
@@ -0,0 +1,177 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.crashlytics.internal.model;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.RandomAccess;
+
+/** Immutable list implementation for use with AutoValue */
+public final class ImmutableList<E> implements List<E>, RandomAccess {
```

Comment:
Did you write this from scratch just for this use case?

Why not just use Collections.unmodifiableList directly and call it good enough?

Reply:
`Collections.unmodifiableList` does not actually expose a type -- it just throws `RuntimeException`s if you try to modify it. At compile-time, it's just a `List`. For AutoValue's setter's, I needed an actual type that would guarantee immutability: https://github.com/google/auto/blob/master/value/userguide/howto.md#mutable_property

Reply:
But this class implements List<E> and has:

```
  public E set(int index, @NonNull E element) {
    return immutableList.set(index, element);
  }
```

So how is that any different than using `Collections.unmodifiableList` directly?

Reply:
This type guarantees that the `List` passed into any of the `CrashlyticsReport` object setters is immutable.

Otherwise the list setters would look like (e.g.): `setEvents(List<Event>)` and would rely on callers to be sure and wrap their mutable lists with `Collections.unmodifiableList`. If they did not, the data object would accidentally be mutable.

Reply:
Essentially the goal was to have a simple `ImmutableList` concrete type without pulling in Guava.

Reply:
I see. I guess that makes sense. And I see the other uses of `ImmutableList` in Firebase are pulled in from Guava for tests, so there doesn't appear to be anything we can easily re-use. So, did you write this class from scratch?

Reply:
Not really. Just created a new `ImmutableList<T>` with an underlying `Collections.unmodifiableList` property, then used `Generate->Delegate methods` in Android Studio. :)